# Imports

In [31]:
import os

import numpy as np
import pandas as pd

import matplotlib.pyplot as plt
import seaborn as sns
sns.set_style('darkgrid', {'legend.frameon':True})

import plotly.graph_objects as go

# Loads

## Old (2015)

### Load

In [32]:
old = pd.read_csv('old/pyla/csv/2015_means.csv', sep=' ', encoding='utf-8')

### Format

Need to reformat date to get same kind of dataset

['date'] -> ['cam', 'year', 'month', 'day', 'hour', 'min']

In [33]:
old.insert(loc=0, column='min', value=old['date'].apply(lambda x:int(x[14:])))
old.insert(loc=0, column='hour', value=old['date'].apply(lambda x:int(x[11:13])))
old.insert(loc=0, column='day', value=old['date'].apply(lambda x:int(x[8:10])))
old.insert(loc=0, column='month', value=old['date'].apply(lambda x:int(x[5:7])))
old.insert(loc=0, column='year', value=old['date'].apply(lambda x:int(x[2:4])))
old.insert(loc=0, column='cam', value=old['date'].apply(lambda x:int(x[0])))

### Rename

In [34]:
old = old.rename(columns={
    'R':'r','G':'g','B':'b',
    'H':'h','S':'s','L':'v',
    'date':'name', 'Tot':'tot',
    'Rpc':'r_pc','Gpc':'g_pc','Bpc':'b_pc',
    'Rex':'r_ex','Gex':'g_ex','Bex':'b_ex',
    'nRGd':'n_rg_d','nRBd':'n_rb_d','nGBd':'n_gb_d'})

### Show

In [35]:
old.info()
old.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2182 entries, 0 to 2181
Data columns (total 23 columns):
cam       2182 non-null int64
year      2182 non-null int64
month     2182 non-null int64
day       2182 non-null int64
hour      2182 non-null int64
min       2182 non-null int64
name      2182 non-null object
r         2182 non-null float64
g         2182 non-null float64
b         2182 non-null float64
h         2182 non-null float64
s         2182 non-null float64
v         2182 non-null float64
tot       2182 non-null float64
r_pc      2182 non-null float64
g_pc      2182 non-null float64
b_pc      2182 non-null float64
r_ex      2182 non-null float64
g_ex      2182 non-null float64
b_ex      2182 non-null float64
n_rg_d    2182 non-null float64
n_rb_d    2182 non-null float64
n_gb_d    2182 non-null float64
dtypes: float64(16), int64(6), object(1)
memory usage: 392.2+ KB


cam  year  month  day  hour  min              name           r           g  \
0    1    15      1    1    10   55  1_15_01_01_10_55  134.300964  132.907079   
1    1    15      1    1    13   55  1_15_01_01_13_55  134.453116  131.764466   
2    1    15      1    1    16   55  1_15_01_01_16_55  137.098643  129.265992   
3    1    15      1    2    10   55  1_15_01_02_10_55  135.171867  133.430806   
4    1    15      1    2    13   55  1_15_01_02_13_55  133.139746  129.243446   

            b  ...         tot      r_pc      g_pc      b_pc       r_ex  \
0  123.558560  ...  390.766603  0.346765  0.341372  0.311863  12.136290   
1  126.594002  ...  392.811584  0.347229  0.336912  0.315859  10.547765   
2  129.730564  ...  396.095199  0.349778  0.327036  0.323184  15.200730   
3  122.194181  ...  390.796855  0.349751  0.342994  0.307255  14.718747   
4  125.067408  ...  387.450601  0.348920  0.334991  0.316088  11.968638   

       g_ex       b_ex    n_rg_d    n_rb_d    n_gb_d  
0  7.954634 -20.090924  0.007504  0.053250  0.045838  
1  2.481813 -13.029578  0.014397  0.047851  0.033610  
2 -8.297224  -6.903506  0.033268  0.039903  0.006735  
3  9.495564 -24.214311  0.009338  0.065161  0.055965  
4  0.279738 -12.248376  0.019570  0.050124  0.030798  

[5 rows x 23 columns]

## News (2014, 2015, 2016, 2017, 2018)

### Load

In [36]:
news = {}
for root, dirs, files in os.walk('csv'):
    for f in files:
        news.update({f.split('_')[1].split('.')[0][2:]:pd.read_csv(os.path.join(root, f), sep=';', encoding='utf-8')})  

### Format

Need to reformat date to get same kind of dataset

['cam', 'year', 'month', 'day', 'hour', 'min'] -> ['date']

(we will delete 'date' later on our new dataset but for now we need it to work faster on our dataset)

In [37]:
for df in news.values():
     df.insert(loc=6, column='name', value=df[['cam', 'year', 'month', 'day', 'hour', 'min']].apply(lambda x: '_'.join([str(i).zfill(2) for i in x])[1:], axis=1))

### Show (2015)

In [38]:
news['15'].info()
news['15'].head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1947 entries, 0 to 1946
Data columns (total 23 columns):
cam       1947 non-null int64
year      1947 non-null int64
month     1947 non-null int64
day       1947 non-null int64
hour      1947 non-null int64
min       1947 non-null int64
name      1947 non-null object
r         1947 non-null float64
g         1947 non-null float64
b         1947 non-null float64
h         1947 non-null float64
s         1947 non-null float64
v         1947 non-null int64
tot       1947 non-null float64
r_pc      1947 non-null float64
g_pc      1947 non-null float64
b_pc      1947 non-null float64
r_ex      1947 non-null float64
g_ex      1947 non-null float64
b_ex      1947 non-null float64
n_rg_d    1947 non-null float64
n_rb_d    1947 non-null float64
n_gb_d    1947 non-null float64
dtypes: float64(15), int64(7), object(1)
memory usage: 350.0+ KB


cam  year  month  day  hour  min              name           r           g  \
0    1    15      1    2    16   55  1_15_01_02_16_55  130.314691  131.573034   
1    1    15      1    3    10   55  1_15_01_03_10_55  126.355285  126.739566   
2    1    15      1    3    13   55  1_15_01_03_13_55  124.816457  123.947530   
3    1    15      1    3    16   55  1_15_01_03_16_55   92.863292  101.232004   
4    1    15      1    4    10   55  1_15_01_04_10_55  123.002006  126.596526   

            b  ...         tot      r_pc      g_pc      b_pc       r_ex  \
0  124.517198  ...  386.404924 -1.084830 -1.068954 -1.152261   4.539150   
1  116.896970  ...  369.991820 -1.066649 -1.065220 -1.162524   9.074034   
2  123.902737  ...  372.666724 -1.090348 -1.102724 -1.097311   1.782646   
3  101.555668  ...  295.650964 -1.177976 -1.070281 -1.063210 -17.061088   
4  128.595659  ...  378.194192 -1.130680 -1.097512 -1.074375  -9.188174   

        g_ex       b_ex    n_rg_d    n_rb_d    n_gb_d  
0   8.314179 -12.853329 -0.008338  0.033326  0.041833  
1  10.226877 -19.300911 -0.001616  0.053086  0.054892  
2  -0.824133  -0.958513  0.005921  0.004569 -0.001289  
3   8.045048   9.016040 -0.053247 -0.056740 -0.003514  
4   1.595387   7.592786 -0.016752 -0.027941 -0.011249  

[5 rows x 23 columns]

# Remove old missing

Old 2015 and new 2015 df have length difference of : 

$2182-1947=235$

$\frac{235}{2182}100\approx10.77%$

This is because when i do the first analysis, we check each missing file and we get them from the source.

But we don't need to do that for now (Maybe if model is rlly close to perfection !).

For check difference between new and old 2015. I need to remove the missing image.

## Remove

In [39]:
old = old[old['name'].isin(news['15']['name'])]
old = old.reset_index(drop=True)

## Show

In [40]:
old.info()
old.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1947 entries, 0 to 1946
Data columns (total 23 columns):
cam       1947 non-null int64
year      1947 non-null int64
month     1947 non-null int64
day       1947 non-null int64
hour      1947 non-null int64
min       1947 non-null int64
name      1947 non-null object
r         1947 non-null float64
g         1947 non-null float64
b         1947 non-null float64
h         1947 non-null float64
s         1947 non-null float64
v         1947 non-null float64
tot       1947 non-null float64
r_pc      1947 non-null float64
g_pc      1947 non-null float64
b_pc      1947 non-null float64
r_ex      1947 non-null float64
g_ex      1947 non-null float64
b_ex      1947 non-null float64
n_rg_d    1947 non-null float64
n_rb_d    1947 non-null float64
n_gb_d    1947 non-null float64
dtypes: float64(16), int64(6), object(1)
memory usage: 350.0+ KB


cam  year  month  day  hour  min              name           r           g  \
0    1    15      1    2    16   55  1_15_01_02_16_55  130.303838  131.571754   
1    1    15      1    3    10   55  1_15_01_03_10_55  126.354776  126.748456   
2    1    15      1    3    13   55  1_15_01_03_13_55  124.816297  123.943888   
3    1    15      1    3    16   55  1_15_01_03_16_55   92.876439  101.232084   
4    1    15      1    4    10   55  1_15_01_04_10_55  122.980968  126.601421   

            b  ...         tot      r_pc      g_pc      b_pc       r_ex  \
0  124.519533  ...  386.395125  0.338769  0.344075  0.317134   4.516390   
1  116.892462  ...  369.995694  0.343942  0.345192  0.310661   9.068634   
2  123.900811  ...  372.660996  0.334760  0.331147  0.332400   1.787895   
3  101.544307  ...  295.652830  0.309969  0.343862  0.346166 -17.023513   
4  128.612246  ...  378.194635  0.323357  0.334260  0.342380  -9.251732   

        g_ex       b_ex    n_rg_d    n_rb_d    n_gb_d  
0   8.320137 -12.836526 -0.008582  0.033114  0.041913  
1  10.249673 -19.318307 -0.001757  0.053055  0.055004  
2  -0.829332  -0.958563  0.005866  0.004612 -0.001211  
3   8.043423   8.980090 -0.053618 -0.057049 -0.003396  
4   1.609628   7.642104 -0.016916 -0.028104 -0.011257  

[5 rows x 23 columns]

# Plot difference

In [41]:
feature_names = ['r', 'g', 'b', 'h', 's', 'v', 'tot', 'r_pc', 'g_pc', 'b_pc', 'r_ex', 'g_ex', 'b_ex', 'n_rg_d', 'n_rb_d', 'n_gb_d']

## Difference between each point for old and new (2015)

In [42]:
diffs = pd.DataFrame()
for n in feature_names:
    diffs[n] = np.abs(old[n] - news['15'][n]) / np.abs(old[n])
diffs.info()
diffs.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1947 entries, 0 to 1946
Data columns (total 16 columns):
r         1947 non-null float64
g         1947 non-null float64
b         1947 non-null float64
h         1947 non-null float64
s         1947 non-null float64
v         1947 non-null float64
tot       1947 non-null float64
r_pc      1947 non-null float64
g_pc      1947 non-null float64
b_pc      1947 non-null float64
r_ex      1947 non-null float64
g_ex      1947 non-null float64
b_ex      1947 non-null float64
n_rg_d    1947 non-null float64
n_rb_d    1947 non-null float64
n_gb_d    1947 non-null float64
dtypes: float64(16)
memory usage: 243.5 KB


r             g         b         h         s    v       tot  \
0  0.000083  9.732608e-06  0.000019  0.010050  0.083430  1.0  0.000025   
1  0.000004  7.014104e-05  0.000039  0.001168  0.022092  1.0  0.000010   
2  0.000001  2.938832e-05  0.000016  0.049532  0.097290  1.0  0.000015   
3  0.000142  7.945795e-07  0.000112  0.029387  0.609917  1.0  0.000006   
4  0.000171  3.866257e-05  0.000129  0.004085  0.066839  1.0  0.000001   

       r_pc      g_pc      b_pc      r_ex      g_ex      b_ex    n_rg_d  \
0  4.202268  4.106743  4.633353  0.005039  0.000716  0.001309  0.028332   
1  4.101244  4.085877  4.742104  0.000595  0.002224  0.000901  0.080338   
2  4.257098  4.330013  4.301173  0.002936  0.006269  0.000052  0.009400   
3  4.800295  4.112526  4.071392  0.002207  0.000202  0.004003  0.006934   
4  4.496694  4.283411  4.137958  0.006870  0.008847  0.006453  0.009696   

     n_rb_d    n_gb_d  
0  0.006409  0.001909  
1  0.000574  0.002037  
2  0.009320  0.063926  
3  0.005425  0.034808  
4  0.005769  0.000681

## Difference between evolution of each couple of old and new (2015)

In [43]:
old_evos = pd.DataFrame()
new_evos = pd.DataFrame()
evos = pd.DataFrame()
for n in feature_names:
    old_tmp = []
    new_tmp = []    
    for i in range(0, len(old) - 1):
        old_tmp.append(np.abs(old.loc[i, n] - old.loc[i + 1, n]))
        new_tmp.append(np.abs(news['15'].loc[i, n] - news['15'].loc[i + 1, n]))
    old_evos[n] = old_tmp
    new_evos[n] = new_tmp
    evos[n] = np.abs(np.array(old_tmp) - np.array(new_tmp)) / np.array(old_tmp)
evos.info()
evos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1946 entries, 0 to 1945
Data columns (total 16 columns):
r         1946 non-null float64
g         1946 non-null float64
b         1946 non-null float64
h         1946 non-null float64
s         1946 non-null float64
v         1946 non-null float64
tot       1946 non-null float64
r_pc      1946 non-null float64
g_pc      1946 non-null float64
b_pc      1946 non-null float64
r_ex      1946 non-null float64
g_ex      1946 non-null float64
b_ex      1946 non-null float64
n_rg_d    1946 non-null float64
n_rb_d    1946 non-null float64
n_gb_d    1946 non-null float64
dtypes: float64(16)
memory usage: 243.4 KB


r         g         b         h           s    v       tot  \
0  0.002620  0.002109  0.000897  0.068491    0.655987  1.0  0.000834   
1  0.000227  0.004469  0.000368  0.510924    0.111036  1.0  0.003603   
2  0.000417  0.000164  0.000422  0.040215  152.707724  1.0  0.000099   
3  0.001136  0.000190  0.001033  0.140332    0.353448  1.0  0.000017   
4  0.010877  0.008161  0.039093  0.189453    0.785991  1.0  0.000156   

       r_pc      g_pc      b_pc      r_ex      g_ex      b_ex    n_rg_d  \
0  2.514516  2.343984  0.585255  0.003814  0.008727  0.005276  0.014938   
1  1.581000  1.670226  1.999681  0.001463  0.002527  0.000945  0.011289   
2  2.534663  1.551413  1.477354  0.001718  0.000403  0.003612  0.005323   
3  2.532789  1.835764  1.949527  0.013013  0.002466  0.063728  0.005661   
4  2.167305  1.774359  1.926046  0.022054  0.016253  0.019109  0.003330   

     n_rb_d    n_gb_d  
0  0.009114  0.002445  
1  0.001516  0.000615  
2  0.005717  0.018664  
3  0.005092  0.016011  
4  0.017367  0.022611

## Plot means

In [44]:
fig = go.Figure([go.Bar(x=feature_names, y=np.mean(diffs, axis=0))])
fig.update_layout({
    'title':'Mean difference between each point for old and new (2015) and for each feature',
    'xaxis_title':'Features',
    'yaxis_title':'Mean rate of difference : abs(old - new) / abs(old)'
})
fig.show()
fig = go.Figure([go.Bar(x=feature_names, y=np.mean(evos, axis=0))])
fig.update_layout({
    'title':'Mean of difference between evolution of each couple of old and new (2015) and for each feature',
    'xaxis_title':'Features',
    'yaxis_title':'Mean rate of difference : abs(evo(old) - evo(new)) / abs(evo(old))'
})
fig.show()

As we can see n_gb_d has an important evolution difference and a much important scale difference

# Plot feature

## Functions

In [45]:
def plot_feature(feature_name):
    plot_feature_value(feature_name)
    plot_feature_diff(feature_name)
    plot_feature_evo(feature_name)
    
def plot_feature_value(feature_name):
    fig = go.Figure()
    fig.add_trace(go.Scatter(x=old['name'], y=old[feature_name], mode='lines', name='old'))
    fig.add_trace(go.Scatter(x=news['15']['name'], y=news['15'][feature_name], mode='lines', name='new'))
    fig.update_layout({
        'title':'Value of %s for each point for old and new (2015)'%feature_name,
        'xaxis_title':'Name',
        'yaxis_title':'Value of %s'%feature_name
    })
    fig.show()
    
def plot_feature_diff(feature_name):
    fig = go.Figure()
    fig.add_trace(go.Scatter(x=old['name'], y=diffs[feature_name], mode='lines'))
    fig.update_layout({
        'title':'Rate of difference of %s for each point for old and new (2015)'%feature_name,
        'xaxis_title':'Name',
        'yaxis_title':'Rate of difference for %s : abs(old - new) / abs(old)'%feature_name
    })
    fig.show()
    
def plot_feature_evo(feature_name):
    fig = go.Figure()
    fig.add_trace(go.Scatter(x=['%s-%s'%(old.loc[i,'name'], old.loc[i+1,'name']) for i in range(0, len(old[feature_name]) - 1)], y=evos[feature_name], mode='lines'))
    fig.update_layout({
        'title':'Rate of evolution of %s for each couple of point for old and new (2015)'%feature_name,
        'xaxis_title':'Couple',
        'yaxis_title':'Rate of evolution for %s : abs(evo(old) - evo(new)) / abs(evo(old))'%feature_name
    })
    fig.show()

## Plot a r_pc

In [46]:
plot_feature('r_pc')

## Plot s

In [47]:
plot_feature('s')

# Create flower feature

We evaluate flower period with our eyes :
- 2014 : (04/08) -> (05/19)
- 2015 : (04/17) -> (05/30)
- 2016 : (05/04) -> (05/31)
- 2017 : (04/14) -> (05/23)
- 2018 : (04/18) -> (05/27)

In [75]:
def flower(x, start, end):
    if x['month'] == start[0]:
        if x['day'] >= start[1]:
            return 1
    elif x['month'] == end[0]:
        if x['day'] <= end[1]:
            return 1
    elif start[0] < x['month'] < end[0]:
        return 1
    return 0

In [82]:
news['14']['flower'] = news['14'].apply(lambda x: flower(x, (4,8),  (5,19)), axis=1)
news['15']['flower'] = news['15'].apply(lambda x: flower(x, (4,17), (5,30)), axis=1)
news['16']['flower'] = news['16'].apply(lambda x: flower(x, (5,4),  (5,31)), axis=1)
news['17']['flower'] = news['17'].apply(lambda x: flower(x, (4,14), (5,23)), axis=1)
news['18']['flower'] = news['18'].apply(lambda x: flower(x, (4,18), (5,27)), axis=1)

# Merge news

In [86]:
np.array([len(value) for value in news.values()]).sum()

10229

In [84]:
df = pd.concat([value for value in news.values()], ignore_index=True)
df.info()
df.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10229 entries, 0 to 10228
Data columns (total 24 columns):
cam       10229 non-null int64
year      10229 non-null int64
month     10229 non-null int64
day       10229 non-null int64
hour      10229 non-null int64
min       10229 non-null int64
name      10229 non-null object
r         10229 non-null float64
g         10229 non-null float64
b         10229 non-null float64
h         10229 non-null float64
s         10229 non-null float64
v         10229 non-null int64
tot       10229 non-null float64
r_pc      10229 non-null float64
g_pc      10229 non-null float64
b_pc      10229 non-null float64
r_ex      10229 non-null float64
g_ex      10229 non-null float64
b_ex      10229 non-null float64
n_rg_d    10229 non-null float64
n_rb_d    10229 non-null float64
n_gb_d    10229 non-null float64
flower    10229 non-null int64
dtypes: float64(15), int64(8), object(1)
memory usage: 1.9+ MB


cam  year  month  day  hour  min              name           r           g  \
0    1    14      1    1    10   55  1_14_01_01_10_55  128.174959  126.531749   
1    1    14      1    1    13   55  1_14_01_01_13_55  123.428187  125.240100   
2    1    14      1    1    16   55  1_14_01_01_16_55  123.780841  127.910258   
3    1    14      1    2    10   55  1_14_01_02_10_55  127.894464  124.938905   
4    1    14      1    2    13   55  1_14_01_02_13_55  121.836390  124.485751   

            b  ...      r_pc      g_pc      b_pc       r_ex      g_ex  \
0  119.842059  ... -1.067369 -1.086410 -1.145645   9.976110  5.046481   
1  125.477267  ... -1.110681 -1.095568 -1.090422  -3.860992  1.574746   
2  131.250416  ... -1.134678 -1.098173 -1.068370 -11.598992  0.789260   
3  118.655820  ... -1.056506 -1.089937 -1.146259  12.194203  3.327525   
4  128.082397  ... -1.126662 -1.103452 -1.066058  -8.895367 -0.947285   

        b_ex    n_rg_d    n_rb_d    n_gb_d  flower  
0 -15.022591  0.009681  0.041008  0.031543       0  
1   2.286246 -0.007658 -0.010277 -0.002500       0  
2  10.809732 -0.018334 -0.033190 -0.014897       0  
3 -15.521728  0.017318  0.049475  0.032636       0  
4   9.842652 -0.012039 -0.030735 -0.018455       0  

[5 rows x 24 columns]

# Write pyla csv

In [88]:
df.drop(['name'], axis=1).to_csv('csv/pyla.csv', ';', index=False)